tests/subsys/pm/device_runtime/src/main.c

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (c) 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sys/printk.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <ztest.h>
+#include <ksched.h>
+#include <kernel.h>
+#include <pm/pm.h>
+#include <pm/device_runtime.h>
+#include "dummy_driver.h"
+
+#define MAX_TIMES 10
+#define STACKSIZE 1024
+
+/* Semaphore used to synchronize thread A and thread B*/
+K_SEM_DEFINE(sem, 0, 1);
+
+K_THREAD_STACK_DEFINE(threadA_stack, STACKSIZE);
+K_THREAD_STACK_DEFINE(threadB_stack, STACKSIZE);
+
+static const struct device *dev;
+static struct dummy_driver_api *api;
+
+static struct k_thread threadA;
+static struct k_thread threadB;
+
+
+void threadA_func(void *arg1, void *arg2, void *arg3)
+{
+	int ret;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	ret = api->open(dev);
+	zassert_true(ret == 0, "Fail to get device");
+
+	/* Lets allow threadB run */
+	k_sem_give(&sem);
+
+	/* Block waiting for device operation conclude */
+	ret = api->wait(dev);
```

Comment:
then both open_sync() and and open() will have the same behavior .... Again, that is just to test device_runtime api.
If I put wait inside the open() then I'll not be able to test from the application multiple async calls and wait after many requests.

Reply:
And btw, having two drivers one using sync api and the other using the async will screw the test because I won't be able to  mix sync/async calls in the same device.

Reply:
wait should not be part of open(), it will have to be called in the path of pm_device_get() and in device_runtime.c or where-ever the logic is implemented in the pm subsystem, if you have to call it outside of the driver APIs then that indicates  something is wrong with the workflow.

Reply:
it is ... it is used in the sync path. But we expose this API, imagine that a device did an async request and later needs to check if this request finished or want to wait until it finishes ? That's the purpose of this function.

Reply:
https://github.com/ceolin/zephyr/blob/pm/deadlock/subsys/pm/device_runtime.c#L155

Reply:
We need to allow a driver knows when an operation has finished or wait to something to finish in the ASYNC case. This API is for this.
This is used internally in the synchronous path as you can see in the link.

Reply:
my point here is that the thread does not need to know about this and should not be calling wait(), sync vs async and pm handling all happens in the driver, not in the calling thread.

Reply:
Yes, but this is not an application, is a test for that API .... Really, I'm not getting the point of over complicate the test.

Reply:
I'm for instance doing multiple calls that are just to stress the API and not a real use case because this is a test to test that API.

Reply:
ok, that test is the only place we are using all of this, we are also missing documentation for this. What I want to see is how all of this would work outside of a test, so, you could add a new test that illustrates how this would work from an application, this new wait() API is just not clicking for me right now and need to see how it would work without it, in the application side of things.

Reply:
Anas, look this https://github.com/ceolin/zephyr/blob/pm/deadlock/subsys/pm/device_runtime.c#L162
This helper is not only used in this test but also in `pm_device_get_sync` and `pm_device_put_sync`.

The API is documented in https://github.com/ceolin/zephyr/blob/pm/deadlock/include/pm/device_runtime.h#L124

I don't know exactly what is missing, but is useful if a device calls `pm_device_get` and then do other tasks it needs and then want to block until the request (`pm_device_get`) concludes.